File: frontend/src/utils/featureEngineering.js
/**
 * 450-dim feature engineering for gRNA efficiency prediction.
 * JavaScript port of backend/app/services/feature_engineering.py
 *
 * Feature vector layout:
 *   [0:80]    Positional one-hot (20 positions × 4 bases)
 *   [80]      GC content
 *   [81]      Normalised Tm (SantaLucia 1998, full guide)
 *   [82:98]   Dinucleotide frequencies (16 pairs)
 *   [98]      Seed region GC (last 12 bp)
 *   [99]      Poly-T flag (TTTT present)
 *   [100:404] Position-specific dinucleotide one-hot (19 × 16)
 *   [404:420] Upstream context one-hot (4 bp)
 *   [420:444] Downstream context one-hot (6 bp)
 *   [444]     GC clamp (last 4 bp)
 *   [445]     RNA hairpin proxy
 *   [446]     Microhomology (from 30-mer)
 *   [447]     Tm PAM-distal (guide[0:8])
 *   [448]     Tm PAM-proximal 8 bp (guide[12:20])
 *   [449]     Tm full 30-mer context
 */

// SantaLucia 1998 DNA-DNA nearest-neighbor parameters
// [ΔH kcal/mol, ΔS cal/mol·K]
const NN = {
  AA:[-7.9,-22.2], AT:[-7.2,-20.4], TA:[-7.2,-21.3], CA:[-8.5,-22.7],
  GT:[-8.4,-22.4], CT:[-7.8,-21.0], GA:[-8.2,-22.2], CG:[-10.6,-27.2],
  GC:[-9.8,-24.4], GG:[-8.0,-19.9], AC:[-7.8,-21.0], TC:[-7.9,-22.2],
  AG:[-8.2,-22.2], TG:[-8.5,-22.7], TT:[-7.9,-22.2], CC:[-8.0,-19.9],
}
const R_GAS    = 1.987   // cal/mol·K
const CT_CONC  = 250e-9  // 250 nM strand concentration
const TM_MIN   = 40.0
const TM_MAX   = 80.0
const GUIDE_LEN = 20

const BASES    = ['A','C','G','T']
const DINUCS   = BASES.flatMap(a => BASES.map(b => a+b))  // 16 pairs
const BASE_IDX = Object.fromEntries(BASES.map((b,i) => [b,i]))
const DINUC_IDX = Object.fromEntries(DINUCS.map((d,i) => [d,i]))

const COMPLEMENT = {A:'T', T:'A', G:'C', C:'G', N:'N'}


// ── Tm calculation (SantaLucia 1998) ─────────────────────────────────────────

export function nearestNeighborTm(seq, oligoConcNM = 250.0) {
  const s = seq.toUpperCase()
  const n = s.length
  if (n < 2) return 0.0

  let dH = 0.0, dS = 0.0
  for (let i = 0; i < n - 1; i++) {
    const di = s[i] + s[i+1]
    if (di in NN) { dH += NN[di][0]; dS += NN[di][1] }
  }
  // Terminal corrections
  for (const end of [s[0], s[n-1]]) {
    if (end === 'A' || end === 'T') { dH += 2.3; dS += 4.1 }
    else if (end === 'G' || end === 'C') { dH += 0.1; dS -= 2.8 }
  }
  const CT = oligoConcNM * 1e-9
  const dHcal = dH * 1000.0
  const denom = dS + R_GAS * Math.log(CT / 4.0)
  if (Math.abs(denom) < 1e-10) return 0.0
  return dHcal / denom - 273.15
}

function tmNorm(seq) {
  if (!seq || seq.length < 4) return 0.0
  const tm = nearestNeighborTm(seq)
  return Math.min(1.0, Math.max(0.0, (tm - TM_MIN) / (TM_MAX - TM_MIN)))
}


// ── Individual feature functions ──────────────────────────────────────────────

function positionalOnehot(seq) {
  const vec = new Float32Array(GUIDE_LEN * 4)
  for (let i = 0; i < seq.length && i < GUIDE_LEN; i++) {
    const idx = BASE_IDX[seq[i]]
    if (idx !== undefined) vec[i * 4 + idx] = 1.0
  }
  return vec
}

function gcContent(seq) {
  if (!seq.length) return 0.0
  let gc = 0
  for (const b of seq) if (b === 'G' || b === 'C') gc++
  return gc / seq.length
}

function dinucFreq(seq) {
  const vec = new Float32Array(16)
  const n = seq.length - 1
  if (n <= 0) return vec
  for (let i = 0; i < n; i++) {
    const idx = DINUC_IDX[seq[i] + seq[i+1]]
    if (idx !== undefined) vec[idx]++
  }
  for (let i = 0; i < 16; i++) vec[i] /= n
  return vec
}

function positionalDinucOnehot(seq) {
  const vec = new Float32Array(19 * 16)
  for (let i = 0; i < Math.min(19, seq.length - 1); i++) {
    const idx = DINUC_IDX[seq[i] + seq[i+1]]
    if (idx !== undefined) vec[i * 16 + idx] = 1.0
  }
  return vec
}

function contextOnehot(ctx, length) {
  const vec = new Float32Array(length * 4)
  const s = (ctx || '').toUpperCase()
  for (let i = 0; i < Math.min(length, s.length); i++) {
    const idx = BASE_IDX[s[i]]
    if (idx !== undefined) vec[i * 4 + idx] = 1.0
  }
  return vec
}

function gcClamp(seq) {
  const tail = seq.slice(-4).toUpperCase()
  let gc = 0
  for (const b of tail) if (b === 'G' || b === 'C') gc++
  return gc / 4.0
}

function hairpinProxy(seq) {
  const s = seq.toUpperCase()
  const n = s.length
  let maxStem = 0
  for (let i = 0; i < n - 8; i++) {
    for (let stemLen = 2; i + stemLen * 2 + 4 <= n; stemLen++) {
      const arm5 = s.slice(i, i + stemLen)
      const arm3Start = i + stemLen + 4
      const arm3 = s.slice(arm3Start, arm3Start + stemLen)
      // Check if arm3 is reverse complement of arm5
      let rc = ''
      for (let k = arm3.length - 1; k >= 0; k--) rc += COMPLEMENT[arm3[k]] || 'N'
      if (arm5 === rc) maxStem = Math.max(maxStem, stemLen)
    }
  }
  return Math.min(1.0, maxStem / 10.0)
}

function microhomology(thirtyMer) {
  if (!thirtyMer || thirtyMer.length < 27) return 0.0
  const tm = thirtyMer.toUpperCase()
  const cut = 21  // SpCas9 cut at guide pos 17 in the 30-mer (4+17=21)
  const left  = tm.slice(Math.max(0, cut - 6), cut)
  const right = tm.slice(cut, Math.min(tm.length, cut + 6))
  let maxMH = 0
  for (let length = 1; length <= Math.min(left.length, right.length); length++) {
    if (left.slice(-length) === right.slice(0, length)) maxMH = length
  }
  return Math.min(1.0, maxMH / 6.0)
}


// ── Main feature extractor ────────────────────────────────────────────────────

/**
 * Extract 450-dim feature vector for a single gRNA.
 * @param {string} sequence   - 20-bp guide sequence
 * @param {string} thirtyMer  - Optional 30-mer context (4bp up + 20 guide + 6bp down)
 * @returns {Float32Array} length 450
 */
export function extractFeatures(sequence, thirtyMer = '') {
  const seq = sequence.toUpperCase().slice(0, GUIDE_LEN).padEnd(GUIDE_LEN, 'N')

  const onehot   = positionalOnehot(seq)                          // 80
  const gc       = new Float32Array([gcContent(seq)])             // 1
  const tm       = new Float32Array([tmNorm(seq)])                // 1
  const dinuc    = dinucFreq(seq)                                 // 16
  const seedGC   = new Float32Array([gcContent(seq.slice(-12))])  // 1
  const polyT    = new Float32Array([seq.includes('TTTT') ? 1.0 : 0.0]) // 1
  const posDinuc = positionalDinucOnehot(seq)                     // 304

  let upstream, downstream
  if (thirtyMer && thirtyMer.length >= 30) {
    const tm30 = thirtyMer.toUpperCase()
    upstream   = contextOnehot(tm30.slice(0, 4), 4)    // 16
    downstream = contextOnehot(tm30.slice(24), 6)      // 24
  } else {
    upstream   = new Float32Array(16)
    downstream = new Float32Array(24)
  }

  const gcClampV = new Float32Array([gcClamp(seq)])       // 1
  const hairpin  = new Float32Array([hairpinProxy(seq)])  // 1
  const mhScore  = new Float32Array([microhomology(thirtyMer || '')]) // 1

  const tmDistal    = new Float32Array([tmNorm(seq.slice(0, 8))])   // 1
  const tmProximal  = new Float32Array([tmNorm(seq.slice(12))])     // 1
  const tmCtx       = thirtyMer && thirtyMer.length >= 30
    ? new Float32Array([tmNorm(thirtyMer.toUpperCase())])
    : new Float32Array(1)                                            // 1

  // Concatenate all into a single Float32Array of length 450
  const parts = [onehot, gc, tm, dinuc, seedGC, polyT, posDinuc,
                 upstream, downstream, gcClampV, hairpin, mhScore,
                 tmDistal, tmProximal, tmCtx]
  const total = parts.reduce((s, p) => s + p.length, 0)
  const result = new Float32Array(total)
  let offset = 0
  for (const p of parts) { result.set(p, offset); offset += p.length }
  return result  // length 450
}

/**
 * Extract features for a batch of guides.
 * @param {string[]} sequences   - Array of 20-bp guide sequences
 * @param {string[]} thirtyMers  - Parallel array of 30-mer context strings
 * @returns {Float32Array} shape (N × 450), row-major
 */
export function extractFeaturesBatch(sequences, thirtyMers = null) {
  const tms = thirtyMers || sequences.map(() => '')
  const N   = sequences.length
  const result = new Float32Array(N * 450)
  for (let i = 0; i < N; i++) {
    result.set(extractFeatures(sequences[i], tms[i]), i * 450)
  }
  return result
}

File: backend/app/services/feature_engineering.py
"""
Feature engineering for ML-based gRNA efficiency prediction.

Feature vector (452 dimensions total):
  [0:80]    Positional one-hot — 20 positions × 4 bases (guide sequence)
  [80]      GC content fraction (guide)
  [81]      Nearest-neighbor Tm (SantaLucia 1998, full guide), normalised 0-1
  [82:98]   Dinucleotide frequencies — all 16 pairs (guide)
  [98]      Seed region GC (last 12 bp, PAM-proximal)
  [99]      Poly-T flag — 1 if TTTT present (RNA Pol III termination)
  [100:404] Position-specific dinucleotide one-hot — 19 positions × 16 (guide)
  [404:420] 4-bp upstream context one-hot  — positions -4,-3,-2,-1
  [420:444] 6-bp downstream context one-hot — PAM+3 bp after (positions +1…+6)
  [444]     GC clamp — GC fraction in last 4 bp (3' end of guide, PAM-proximal)
  [445]     Tm asymmetry — abs(Tm(guide[0:10]) − Tm(guide[10:20])) / 20, normalised 0-1
  [446]     Microhomology length — max homology at SpCas9 cut site (0-1, from 30-mer)
  [447]     Tm of PAM-distal half (guide positions 0:8), normalised 0-1
  [448]     Seed region ΔG — duplex free energy of guide[12:20] at 37°C, normalised 0-1
  [449]     Tm of full 30-mer context, normalised 0-1 (0 if 30-mer unavailable)
  [450]     PAM-proximal 10bp GC (positions 11–20, non-template strand)
  [451]     PAM-distal 10bp GC (positions 1–10, non-template strand)

When a 30-mer is not available (Doench 2014 guides), dims [404:420], [420:444],
[446], and [449] are zero-padded.

Biological references:
  - One-hot positional encoding:    Doench et al. 2016 Nat Biotechnol 34:184
  - Nearest-neighbor Tm:            SantaLucia 1998 PNAS 95:1460
  - Dinucleotide contributions:     Doench 2016 Supplementary Table S9
  - Seed region GC:                 Hsu et al. 2013 Nat Biotech 31:827
  - Poly-T flag:                    Brummelkamp et al. 2002 Science 296:550
  - Position-specific dinucs:       Doench 2016 Azimuth feature set
  - Flanking context:               Doench 2016 (30-mer); Kim et al. 2019
  - GC clamp (3' end):              Doench et al. 2016 (3'-end GC composition)
  - Tm asymmetry (5'/3' halves):    proxy for guide folding asymmetry; replaces hairpin heuristic
  - Microhomology:                  Bae et al. 2014 Genome Research 24:132
  - Segmented Tm windows:           Doench 2016 Azimuth (4 Tm sub-windows)
"""
import numpy as np
from typing import List, Optional
from app.utils.biology_utils import nearest_neighbor_tm, seed_region_gc, has_poly_t, seed_delta_g

BASES     = ["A", "C", "G", "T"]
DINUCS    = [a + b for a in BASES for b in BASES]   # 16 dinucleotides
GUIDE_LEN = 20
N_FEATURES = 452   # 80+1+1+16+1+1+304+16+24+1+1+1+1+1+1+1+1 (added proximal/distal GC split)

_COMPLEMENT = {"A": "T", "T": "A", "G": "C", "C": "G", "N": "N"}

_TM_MIN = 40.0
_TM_MAX = 80.0
_BASE_IDX = {b: i for i, b in enumerate(BASES)}
_DINUC_IDX = {d: i for i, d in enumerate(DINUCS)}


def _positional_onehot(sequence: str) -> np.ndarray:
    """80-dim one-hot: position i × base b → index i*4 + BASES.index(b)."""
    vec = np.zeros(GUIDE_LEN * 4, dtype=np.float32)
    for i, base in enumerate(sequence.upper()):
        idx = _BASE_IDX.get(base)
        if idx is not None:
            vec[i * 4 + idx] = 1.0
    return vec


def _gc_content(sequence: str) -> float:
    seq = sequence.upper()
    return (seq.count("G") + seq.count("C")) / len(seq) if seq else 0.0


def _tm_normalised(sequence: str) -> float:
    tm = nearest_neighbor_tm(sequence)
    return float(np.clip((tm - _TM_MIN) / (_TM_MAX - _TM_MIN), 0.0, 1.0))


def _dinucleotide_freq(sequence: str) -> np.ndarray:
    """16-dim normalised dinucleotide frequency vector."""
    seq = sequence.upper()
    vec = np.zeros(16, dtype=np.float32)
    n   = len(seq) - 1
    if n <= 0:
        return vec
    for i in range(n):
        idx = _DINUC_IDX.get(seq[i:i + 2])
        if idx is not None:
            vec[idx] += 1
    vec /= n
    return vec


def _positional_dinucleotide_onehot(sequence: str) -> np.ndarray:
    """304-dim position-specific dinucleotide one-hot (Doench 2016 / Azimuth)."""
    seq = sequence.upper()
    vec = np.zeros(19 * 16, dtype=np.float32)
    for i in range(min(19, len(seq) - 1)):
        idx = _DINUC_IDX.get(seq[i:i + 2])
        if idx is not None:
            vec[i * 16 + idx] = 1.0
    return vec


def _context_onehot(context: str, length: int) -> np.ndarray:
    """One-hot encode an upstream or downstream context window (4×length dims)."""
    vec = np.zeros(length * 4, dtype=np.float32)
    ctx = context.upper()
    for i, base in enumerate(ctx[:length]):
        idx = _BASE_IDX.get(base)
        if idx is not None:
            vec[i * 4 + idx] = 1.0
    return vec


def _tm_segment(segment: str) -> float:
    """Normalised Tm for an arbitrary sub-sequence (SantaLucia 1998).

    Used for the three segmented Tm windows inspired by Azimuth's feature set:
      - PAM-distal  (guide positions  0:8)
      - Seed region (guide positions 12:20)
      - Full 30-mer context (when available)
    Returns 0.0 for segments shorter than 4 bases.
    """
    if len(segment) < 2:
        return 0.0
    tm = nearest_neighbor_tm(segment)
    return float(np.clip((tm - _TM_MIN) / (_TM_MAX - _TM_MIN), 0.0, 1.0))


def _proximal_gc(sequence: str) -> float:
    """GC fraction in PAM-proximal 10 bp (positions 11–20, non-template strand).
    Captures R-loop stability near the PAM where Cas9 initiates strand invasion."""
    seg = sequence.upper()[10:20]
    return (seg.count("G") + seg.count("C")) / max(len(seg), 1)


def _distal_gc(sequence: str) -> float:
    """GC fraction in PAM-distal 10 bp (positions 1–10, non-template strand).
    Elevated distal GC can increase off-target hybridisation at secondary sites."""
    seg = sequence.upper()[:10]
    return (seg.count("G") + seg.count("C")) / max(len(seg), 1)


def _gc_clamp(sequence: str) -> float:
    """GC fraction in the last 4 positions (3' end, PAM-proximal).

    Guides with 2-3 GC in the last 4 bp tend to have higher cleavage efficiency
    (Doench et al. 2016 positional analysis).  Returns 0-1.
    """
    tail = sequence.upper()[-4:]
    return (tail.count("G") + tail.count("C")) / 4.0


def _tm_asymmetry(sequence: str) -> float:
    """Thermodynamic asymmetry between 5' and 3' halves of the guide.

    abs(Tm(guide[0:10]) − Tm(guide[10:20])) normalised by 20°C practical max.
    Guides with a cool 5' end and warm 3' (seed) end show better R-loop formation
    and cleavage efficiency.  More informative than binary hairpin detection.
    Normalised to [0, 1].
    """
    tm_5 = nearest_neighbor_tm(sequence[:10])
    tm_3 = nearest_neighbor_tm(sequence[10:20])
    return float(np.clip(abs(tm_5 - tm_3) / 20.0, 0.0, 1.0))


def _microhomology(thirty_mer: str) -> float:
    """Microhomology length at the SpCas9 cut site, normalised to [0, 1].

    SpCas9 cuts between positions 17 and 18 of the 20-bp guide (3 bp upstream
    of the NGG PAM).  In the 30-mer context (4-bp upstream + 20-bp guide +
    6-bp downstream), the cut is at index 4+17 = 21.

    Microhomology: identical bases directly flanking the cut (Bae et al. 2014).
    Up to 6 bp checked on each side.  Score = longest matching prefix / 6.
    """
    if not thirty_mer or len(thirty_mer) < 27:
        return 0.0
    tm = thirty_mer.upper()
    cut = 21  # position after cut in 30-mer
    left  = tm[max(0, cut - 6): cut]       # up to 6 bp left of cut
    right = tm[cut: min(len(tm), cut + 6)] # up to 6 bp right of cut
    # Count matching bases from the cut outward
    max_mh = 0
    for length in range(1, min(len(left), len(right)) + 1):
        if left[-length:] == right[:length]:
            max_mh = length
    return min(1.0, max_mh / 6.0)


def extract_features(sequence: str,
                     thirty_mer: Optional[str] = None) -> np.ndarray:
    """
    Extract 450-dim feature vector from a 20-bp guide and optional 30-mer context.

    Args:
        sequence:   20-bp guide sequence (5'→3')
        thirty_mer: Full 30-mer (4bp upstream + 20bp guide + 6bp downstream).
                    When provided, upstream/downstream features are populated.
                    When None or empty, those 40 dims are zero-padded.

    Returns:
        np.ndarray of shape (450,)
    """
    seq = sequence.upper()[:GUIDE_LEN].ljust(GUIDE_LEN, "N")

    onehot    = _positional_onehot(seq)                              # 80
    gc        = np.array([_gc_content(seq)],  dtype=np.float32)     # 1
    tm        = np.array([_tm_normalised(seq)], dtype=np.float32)   # 1
    dinuc     = _dinucleotide_freq(seq)                              # 16
    seed_gc   = np.array([seed_region_gc(seq)], dtype=np.float32)   # 1
    poly_t    = np.array([1.0 if has_poly_t(seq, 4) else 0.0],
                         dtype=np.float32)                           # 1
    pos_dinuc = _positional_dinucleotide_onehot(seq)                 # 304

    # Flanking context — zero if not available
    if thirty_mer and len(thirty_mer) >= 30:
        tm30 = thirty_mer.upper()
        upstream   = _context_onehot(tm30[:4],  length=4)           # 16
        downstream = _context_onehot(tm30[24:], length=6)           # 24
    else:
        upstream   = np.zeros(16, dtype=np.float32)                 # 16
        downstream = np.zeros(24, dtype=np.float32)                 # 24

    # Advanced biological features
    gc_clamp    = np.array([_gc_clamp(seq)],                  dtype=np.float32)  # 1
    tm_asym     = np.array([_tm_asymmetry(seq)],              dtype=np.float32)  # 1  dim 445
    mh_score    = np.array([_microhomology(thirty_mer or "")], dtype=np.float32)  # 1

    # Segmented Tm / thermodynamic windows:
    #   dim 447: PAM-distal Tm (guide[0:8])
    #   dim 448: seed region ΔG at 37°C (guide[12:20]), normalised 0-1
    #            more negative ΔG = more stable seed duplex → higher value
    #   dim 449: full 30-mer Tm context (0 if unavailable)
    tm_pam_distal = np.array([_tm_segment(seq[:8])], dtype=np.float32)           # 1

    # Normalise seed ΔG: practical range ~[-4.5, 0.5] kcal/mol → map to [0,1]
    _DG_MIN, _DG_MAX = -4.5, 0.5
    raw_dg = seed_delta_g(seq)
    seed_dg = np.array([float(np.clip((raw_dg - _DG_MAX) / (_DG_MIN - _DG_MAX), 0.0, 1.0))],
                        dtype=np.float32)                                          # 1

    if thirty_mer and len(thirty_mer) >= 30:
        tm_ctx = np.array([_tm_segment(thirty_mer.upper())], dtype=np.float32)   # 1
    else:
        tm_ctx = np.zeros(1, dtype=np.float32)                                   # 1

    prox_gc   = np.array([_proximal_gc(seq)],  dtype=np.float32)                 # 1
    distal_gc = np.array([_distal_gc(seq)],    dtype=np.float32)                 # 1

    return np.concatenate(
        [onehot, gc, tm, dinuc, seed_gc, poly_t, pos_dinuc,
         upstream, downstream, gc_clamp, tm_asym, mh_score,
         tm_pam_distal, seed_dg, tm_ctx,
         prox_gc, distal_gc]
    )   # 452


def extract_features_batch(sequences: List[str],
                            thirty_mers: Optional[List[str]] = None
                            ) -> np.ndarray:
    """
    Extract features for a list of sequences.

    Args:
        sequences:   List of 20-bp guide sequences
        thirty_mers: Optional parallel list of 30-mer context strings.
                     Pass None to use guide-only features for all.

    Returns:
        np.ndarray of shape (N, 450)
    """
    if thirty_mers is None:
        thirty_mers = [""] * len(sequences)
    return np.vstack([
        extract_features(s, t) for s, t in zip(sequences, thirty_mers)
    ])
